// @stencil/router/dist/collection/components/__tests__/test-demo-seven.js
export class TestDemoSeven {
    componentWillLoad() {
        console.log(`%c test-demo-seven - componentWillLoad`, 'color: red');
    }
    componentDidLoad() {
        console.log(`%c test-demo-seven - componentDidLoad`, 'color: orange');
    }
    componentWillUpdate() {
        console.log(`%c test-demo-seven - componentWillUpdate`, 'color: green');
    }
    componentDidUpdate() {
        console.log(`%c test-demo-seven - componentDidUpdate`, 'color: blue');
    }
    render() {
        return (h("div", null,
            h("h1", null, "Test Demo Seven"),
            h("pre", null,
                h("b", null, "this.pages"),
                ":",
                h("br", null),
                JSON.stringify(this.pages, null, 2)),
            h("pre", null,
                h("b", null, "this.match"),
                ":",
                h("br", null),
                JSON.stringify(this.match, null, 2))));
    }
    static get is() { return "test-demo-seven"; }
    static get properties() { return {
        "history": {
            "type": "Any",
            "attr": "history"
        },
        "match": {
            "type": "Any",
            "attr": "match"
        },
        "pages": {
            "type": "Any",
            "attr": "pages"
        }
    }; }
}
