// @stencil/router/dist/esm/es5/chunk-4b7cc8ca.js
var DEFAULT_DELIMITER = "/", DEFAULT_DELIMITERS = "./", PATH_REGEXP = new RegExp(["(\\\\.)", "(?:\\:(\\w+)(?:\\(((?:\\\\.|[^\\\\()])+)\\))?|\\(((?:\\\\.|[^\\\\()])+)\\))([+*?])?"].join("|"), "g");
function parse(e, t) { for (var n, a = [], r = 0, s = 0, i = "", o = t && t.delimiter || DEFAULT_DELIMITER, c = t && t.delimiters || DEFAULT_DELIMITERS, u = !1; null !== (n = PATH_REGEXP.exec(e));) {
    var h = n[0], l = n[1], p = n.index;
    if (i += e.slice(s, p), s = p + h.length, l)
        i += l[1], u = !0;
    else {
        var f = "", d = e[s], g = n[2], m = n[3], E = n[4], v = n[5];
        if (!u && i.length) {
            var x = i.length - 1;
            c.indexOf(i[x]) > -1 && (f = i[x], i = i.slice(0, x));
        }
        i && (a.push(i), i = "", u = !1);
        var A = "" !== f && void 0 !== d && d !== f, y = "+" === v || "*" === v, R = "?" === v || "*" === v, O = f || o, T = m || E;
        a.push({ name: g || r++, prefix: f, delimiter: O, optional: R, repeat: y, partial: A, pattern: T ? escapeGroup(T) : "[^" + escapeString(O) + "]+?" });
    }
} return (i || s < e.length) && a.push(i + e.substr(s)), a; }
function escapeString(e) { return e.replace(/([.+*?=^!:${}()[\]|/\\])/g, "\\$1"); }
function escapeGroup(e) { return e.replace(/([=!:$/()])/g, "\\$1"); }
function flags(e) { return e && e.sensitive ? "" : "i"; }
function regexpToRegexp(e, t) { if (!t)
    return e; var n = e.source.match(/\((?!\?)/g); if (n)
    for (var a = 0; a < n.length; a++)
        t.push({ name: a, prefix: null, delimiter: null, optional: !1, repeat: !1, partial: !1, pattern: null }); return e; }
function arrayToRegexp(e, t, n) { for (var a = [], r = 0; r < e.length; r++)
    a.push(pathToRegexp(e[r], t, n).source); return new RegExp("(?:" + a.join("|") + ")", flags(n)); }
function stringToRegexp(e, t, n) { return tokensToRegExp(parse(e, n), t, n); }
function tokensToRegExp(e, t, n) { for (var a = (n = n || {}).strict, r = !1 !== n.end, s = escapeString(n.delimiter || DEFAULT_DELIMITER), i = n.delimiters || DEFAULT_DELIMITERS, o = [].concat(n.endsWith || []).map(escapeString).concat("$").join("|"), c = "", u = !1, h = 0; h < e.length; h++) {
    var l = e[h];
    if ("string" == typeof l)
        c += escapeString(l), u = h === e.length - 1 && i.indexOf(l[l.length - 1]) > -1;
    else {
        var p = escapeString(l.prefix), f = l.repeat ? "(?:" + l.pattern + ")(?:" + p + "(?:" + l.pattern + "))*" : l.pattern;
        t && t.push(l), l.optional ? l.partial ? c += p + "(" + f + ")?" : c += "(?:" + p + "(" + f + "))?" : c += p + "(" + f + ")";
    }
} return r ? (a || (c += "(?:" + s + ")?"), c += "$" === o ? "$" : "(?=" + o + ")") : (a || (c += "(?:" + s + "(?=" + o + "))?"), u || (c += "(?=" + s + "|" + o + ")")), new RegExp("^" + c, flags(n)); }
function pathToRegexp(e, t, n) { return e instanceof RegExp ? regexpToRegexp(e, t) : Array.isArray(e) ? arrayToRegexp(e, t, n) : stringToRegexp(e, t, n); }
function hasBasename(e, t) { return new RegExp("^" + t + "(\\/|\\?|#|$)", "i").test(e); }
function stripBasename(e, t) { return hasBasename(e, t) ? e.substr(t.length) : e; }
function stripTrailingSlash(e) { return "/" === e.charAt(e.length - 1) ? e.slice(0, -1) : e; }
function addLeadingSlash(e) { return "/" === e.charAt(0) ? e : "/" + e; }
function stripLeadingSlash(e) { return "/" === e.charAt(0) ? e.substr(1) : e; }
function parsePath(e) { var t = e || "/", n = "", a = ""; var r = t.indexOf("#"); -1 !== r && (a = t.substr(r), t = t.substr(0, r)); var s = t.indexOf("?"); return -1 !== s && (n = t.substr(s), t = t.substr(0, s)), { pathname: t, search: "?" === n ? "" : n, hash: "#" === a ? "" : a }; }
function createPath(e) { var t = e.pathname, n = e.search, a = e.hash; var r = t || "/"; return n && "?" !== n && (r += "?" === n.charAt(0) ? n : "?" + n), a && "#" !== a && (r += "#" === a.charAt(0) ? a : "#" + a), r; }
function parseQueryString(e) { return e ? (/^[?#]/.test(e) ? e.slice(1) : e).split("&").reduce(function (e, t) { var _a = t.split("="), n = _a[0], a = _a[1]; return e[n] = a ? decodeURIComponent(a.replace(/\+/g, " ")) : "", e; }, {}) : {}; }
function isAbsolute(e) { return "/" === e.charAt(0); }
function spliceOne(e, t) { for (var n = t, a = n + 1, r = e.length; a < r; n += 1, a += 1)
    e[n] = e[a]; e.pop(); }
function resolvePathname(e, t) {
    if (t === void 0) { t = ""; }
    var n = e && e.split("/") || [];
    var a = t && t.split("/") || [];
    var r = e && isAbsolute(e), s = t && isAbsolute(t), i = r || s;
    if (e && isAbsolute(e) ? a = n : n.length && (a.pop(), a = a.concat(n)), !a.length)
        return "/";
    var o;
    if (a.length) {
        var e_1 = a[a.length - 1];
        o = "." === e_1 || ".." === e_1 || "" === e_1;
    }
    else
        o = !1;
    var c = 0;
    for (var e_2 = a.length; e_2 >= 0; e_2--) {
        var t_1 = a[e_2];
        "." === t_1 ? spliceOne(a, e_2) : ".." === t_1 ? (spliceOne(a, e_2), c++) : c && (spliceOne(a, e_2), c--);
    }
    if (!i)
        for (; c--; c)
            a.unshift("..");
    !i || "" === a[0] || a[0] && isAbsolute(a[0]) || a.unshift("");
    var u = a.join("/");
    return o && "/" !== u.substr(-1) && (u += "/"), u;
}
function valueEqual(e, t) { if (e === t)
    return !0; if (null == e || null == t)
    return !1; if (Array.isArray(e))
    return Array.isArray(t) && e.length === t.length && e.every(function (e, n) { return valueEqual(e, t[n]); }); var n = typeof e; if (n !== typeof t)
    return !1; if ("object" === n) {
    var n_1 = e.valueOf(), a = t.valueOf();
    if (n_1 !== e || a !== t)
        return valueEqual(n_1, a);
    var r = Object.keys(e), s = Object.keys(t);
    return r.length === s.length && r.every(function (n) { return valueEqual(e[n], t[n]); });
} return !1; }
function locationsAreEqual(e, t) { return e.pathname === t.pathname && e.search === t.search && e.hash === t.hash && e.key === t.key && valueEqual(e.state, t.state); }
function createLocation(e, t, n, a) { var r; "string" == typeof e ? (r = parsePath(e)).state = t : (void 0 === (r = Object.assign({}, e)).pathname && (r.pathname = ""), r.search ? "?" !== r.search.charAt(0) && (r.search = "?" + r.search) : r.search = "", r.hash ? "#" !== r.hash.charAt(0) && (r.hash = "#" + r.hash) : r.hash = "", void 0 !== t && void 0 === r.state && (r.state = t)); try {
    r.pathname = decodeURI(r.pathname);
}
catch (e) {
    throw e instanceof URIError ? new URIError('Pathname "' + r.pathname + '" could not be decoded. This is likely caused by an invalid percent-encoding.') : e;
} return n && (r.key = n), a ? r.pathname ? "/" !== r.pathname.charAt(0) && (r.pathname = resolvePathname(r.pathname, a.pathname)) : r.pathname = a.pathname : r.pathname || (r.pathname = "/"), r.query = parseQueryString(r.search), r; }
var patternCache = {}, cacheLimit = 1e4;
var cacheCount = 0;
function compilePath(e, t) { var n = "" + t.end + t.strict, a = patternCache[n] || (patternCache[n] = {}), r = JSON.stringify(e); if (a[r])
    return a[r]; var s = [], i = { re: pathToRegexp(e, s, t), keys: s }; return cacheCount < cacheLimit && (a[r] = i, cacheCount += 1), i; }
function matchPath(e, t) {
    if (t === void 0) { t = {}; }
    "string" == typeof t && (t = { path: t });
    var _a = t.path, n = _a === void 0 ? "/" : _a, _b = t.exact, a = _b === void 0 ? !1 : _b, _c = t.strict, r = _c === void 0 ? !1 : _c, _d = compilePath(n, { end: a, strict: r }), s = _d.re, i = _d.keys, o = s.exec(e);
    if (!o)
        return null;
    var c = o[0], u = o.slice(1), h = e === c;
    return a && !h ? null : { path: n, url: "/" === n && "" === c ? "/" : c, isExact: h, params: i.reduce(function (e, t, n) { return e[t.name] = u[n], e; }, {}) };
}
function matchesAreEqual(e, t) { return null == e && null == t || null != t && e && t && e.path === t.path && e.url === t.url && valueEqual(e.params, t.params); }
var canUseDOM = !("undefined" == typeof window || !window.document || !window.document.createElement), addEventListener = function (e, t, n) { return e.addEventListener ? e.addEventListener(t, n, !1) : e.attachEvent("on" + t, n); }, removeEventListener = function (e, t, n) { return e.removeEventListener ? e.removeEventListener(t, n, !1) : e.detachEvent("on" + t, n); }, getConfirmation = function (e, t) { return t(window.confirm(e)); }, isModifiedEvent = function (e) { return e.metaKey || e.altKey || e.ctrlKey || e.shiftKey; }, supportsHistory = function () { var e = window.navigator.userAgent; return (-1 === e.indexOf("Android 2.") && -1 === e.indexOf("Android 4.0") || -1 === e.indexOf("Mobile Safari") || -1 !== e.indexOf("Chrome") || -1 !== e.indexOf("Windows Phone")) && window.history && "pushState" in window.history; }, supportsPopStateOnHashChange = function () { return -1 === window.navigator.userAgent.indexOf("Trident"); }, supportsGoWithoutReloadUsingHash = function () { return -1 === window.navigator.userAgent.indexOf("Firefox"); }, isExtraneousPopstateEvent = function (e) { return void 0 === e.state && -1 === navigator.userAgent.indexOf("CriOS"); }, storageAvailable = function (e) { try {
    var t = window[e], n = "__storage_test__";
    return t.setItem(n, n), t.removeItem(n), !0;
}
catch (e) {
    return e instanceof DOMException && (22 === e.code || 1014 === e.code || "QuotaExceededError" === e.name || "NS_ERROR_DOM_QUOTA_REACHED" === e.name) && 0 !== t.length;
} };
export { matchPath as a, matchesAreEqual as b, isModifiedEvent as c, storageAvailable as d, createLocation as e, addLeadingSlash as f, stripTrailingSlash as g, hasBasename as h, stripBasename as i, createPath as j, canUseDOM as k, addEventListener as l, removeEventListener as m, getConfirmation as n, supportsHistory as o, supportsPopStateOnHashChange as p, isExtraneousPopstateEvent as q, locationsAreEqual as r, stripLeadingSlash as s, supportsGoWithoutReloadUsingHash as t };
